File: src/dr_stone/failures.py
from __future__ import annotations

from dr_stone.exceptions import FetchError, ParseError
from dr_stone.models import ScrapeFailure


def build_scrape_failure(source: str, target_url: str, exc: Exception) -> ScrapeFailure:
    if isinstance(exc, FetchError):
        return ScrapeFailure(
            source=source,
            stage="fetch",
            error_code=exc.code,
            error_type=type(exc).__name__,
            message=str(exc),
            target_url=exc.url or target_url,
            retriable=exc.retriable,
            http_status=exc.status_code,
            final_url=exc.final_url,
            details=exc.details,
        )

    if isinstance(exc, ParseError):
        return ScrapeFailure(
            source=source,
            stage="parse",
            error_code=exc.code,
            error_type=type(exc).__name__,
            message=str(exc),
            target_url=target_url,
            retriable=False,
            details=exc.details,
        )

    return ScrapeFailure(
        source=source,
        stage="unknown",
        error_code="unexpected_error",
        error_type=type(exc).__name__,
        message=str(exc),
        target_url=target_url,
        retriable=False,
    )

File: dr-stone-scrapper/src/failures.ts
import type { ScrapeFailure } from "@dr-stone/database";

import { FetchError, ParseError } from "./errors.js";

export function buildScrapeFailure(
  source: string,
  targetUrl: string,
  error: unknown
): ScrapeFailure {
  if (error instanceof FetchError) {
    return {
      source,
      stage: "fetch",
      errorCode: error.code,
      errorType: error.constructor.name,
      message: error.message,
      targetUrl: error.url ?? targetUrl,
      retriable: error.retriable,
      httpStatus: error.statusCode ?? null,
      finalUrl: error.finalUrl ?? null,
      details: error.details,
      capturedAt: new Date().toISOString()
    };
  }

  if (error instanceof ParseError) {
    return {
      source,
      stage: "parse",
      errorCode: error.code,
      errorType: error.constructor.name,
      message: error.message,
      targetUrl,
      retriable: false,
      details: error.details,
      capturedAt: new Date().toISOString()
    };
  }

  return {
    source,
    stage: "unknown",
    errorCode: "unexpected_error",
    errorType: error instanceof Error ? error.constructor.name : "UnknownError",
    message: error instanceof Error ? error.message : String(error),
    targetUrl,
    retriable: false,
    details: {},
    capturedAt: new Date().toISOString()
  };
}
